### PR TITLE
Adding symmetry-aware tensors

### PR DESCRIPTION
**Pull request by cgbriggs99**
Adding facilities to include symmetry-aware tensors. Both point-group and index permutation symmetries may eventually be supported.

### CODE REVIEW

**Comment by jturney** [comment]
Oh boy. The fun begins. 😄

**Comment by cgbriggs99** [comment]
Last commit was a big one. Here's what changed.

- Moved tensor functions out of the gpu namespaces. There is still `einsums::gpu` for GPU utility functions, but now there is no `einsums::tensor_algebra::gpu`, for instance.
- Changed stream management. Now, each OMP thread gets a stream so that things can run more smoothly.
- Minor and major adjustments to things that I found while working on the plugins.

**Comment by cgbriggs99** [comment]
Bit of a change. I'm removing index symmetry from this pull request. The block-diagonal tensors are a big enough change on their own. The index symmetry and tiled tensors will need to be separate pull requests.

File: src/include/einsums/BlockTensor.hpp [hunk 1]
```diff
@@ -0,0 +1,1023 @@
...
+        // Add the new ranges.
+        if (pos == 0) {
+            _ranges.emplace(_ranges.begin(), 0, value.dim(0));
+        } else {
+            _ranges.emplace(std::next(_ranges.begin(), pos), _ranges[pos - 1][1], _ranges[pos - 1][1] + value.dim(0));
+        }
+
+        for (int i = pos + 1; i < _ranges.size(); i++) {
+            _ranges[i][0] += value.dim(0);
+            _ranges[i][1] += value.dim(0);
```
[the remark this answers is not in the document]
Comment (jturney) [reply src/include/einsums/BlockTensor.hpp:234]: Oh nevermind.

File: CMakeLists.txt [hunk 2]
```diff
@@ -69,6 +69,7 @@ einsums_option(EINSUMS_H5CPP_USE_OMP_ALIGNED_ALLOC BOOL
         "On Mac, use omp_aligned_alloc instead of aligned_alloc. Helpful for old SDK." OFF)
 einsums_option(EINSUMS_BUILD_HIP BOOL "Build HIP GPU code." OFF)
 einsums_option(EINSUMS_COVERAGE BOOL "Enable code coverage checking" OFF)
+einsums_option(EINSUMS_PSI4_CONNECTOR BOOL "Build the Psi4 connector and tensor converter." OFF)
```
Comment (jturney) [inline CMakeLists.txt:72]: I don't think this is needed anymore, is it?

File: src/include/einsums/BlockTensor.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cgbriggs99) [reply src/include/einsums/BlockTensor.hpp:234]: For now, yes. The blocks are assumed to be square.

File: CMakeLists.txt (hunk 2 above)
Comment (cgbriggs99) [reply CMakeLists.txt:72]: Oops.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -45,6 +45,14 @@ extend_einsums_target(einsums
         hip::device
 )
 
+extend_einsums_target(einsums
```
Comment (jturney) [inline src/CMakeLists.txt:48]: Or these lines.

**Pull request merged**